File: java/base/src/main/vsdk/toolkit/environment/light/SpotLight.java
package vsdk.toolkit.environment.light;

import vsdk.toolkit.common.VSDK;
import vsdk.toolkit.common.linealAlgebra.Vector3Dd;
import vsdk.toolkit.common.color.ColorRgb;
import vsdk.toolkit.environment.geometry.element.Ray;

public final class SpotLight extends Light
{
    private final Vector3Dd pointsAt;
    private final double coefficient;
    private final double radius;
    private final double falloff;

    public SpotLight(
        Vector3Dd position,
        Vector3Dd pointsAt,
        ColorRgb emission,
        double coefficient,
        double radius,
        double falloff)
    {
        super(position, emission);
        this.pointsAt = pointsAt;
        this.coefficient = coefficient;
        this.radius = radius;
        this.falloff = falloff;
    }

    public Vector3Dd getPointsAt()
    {
        return pointsAt;
    }

    public double getCoefficient()
    {
        return coefficient;
    }

    public double getRadius()
    {
        return radius;
    }

    public double getFalloff()
    {
        return falloff;
    }

    @Override
    public LightDirection getDirectionAndDistance(Vector3Dd surfacePoint)
    {
        Vector3Dd toLight = getPosition().subtract(surfacePoint);
        double distance = toLight.length();
        if ( distance <= VSDK.EPSILON ) {
            return new LightDirection(new Vector3Dd(0, 0, 0), 0.0);
        }
        double invDistance = 1.0 / distance;
        return new LightDirection(
            new Vector3Dd(toLight.x() * invDistance, toLight.y() * invDistance, toLight.z() * invDistance),
            distance - VSDK.EPSILON);
    }

    private static double cubicSpline(double low, double high, double pos)
    {
        if ( pos < low ) {
            return 0.0;
        }
        if ( pos > high ) {
            return 1.0;
        }
        if ( high == low ) {
            return 0.0;
        }

        double t = (pos - low) / (high - low);
        return (3 - 2 * t) * t * t;
    }

    @Override
    public double evaluateLightResponseFactor(Ray lightSourceRay)
    {
        Vector3Dd spotDirection = getPointsAt().subtract(getPosition());
        double len = spotDirection.length();
        if ( len <= 0.0 ) {
            return 0.0;
        }
        spotDirection = new Vector3Dd(
            spotDirection.x() / len, spotDirection.y() / len, spotDirection.z() / len);
        double cosTheta = -lightSourceRay.getDirection().dotProduct(spotDirection);
        if ( cosTheta <= 0.0 ) {
            return 0.0;
        }
        double attenuation = Math.pow(cosTheta, getCoefficient());
        if ( getRadius() > 0.0 ) {
            attenuation *= cubicSpline(getFalloff(), getRadius(), cosTheta);
        }
        return attenuation;
    }

    @Override
    public Light copy()
    {
        SpotLight copy = new SpotLight(
            getPosition(), pointsAt, getEmission(), coefficient, radius, falloff);
        copy.setId(getId());
        copy.setName(getName());
        return copy;
    }
}

File: typescript/base/src/main/vsdk/toolkit/environment/light/SpotLight.ts
import { ColorRgb } from "../../common/color/ColorRgb.js";
import { Vector3Dd } from "../../common/linealAlgebra/Vector3Dd.js";
import { VSDK } from "../../common/VSDK.js";
import type { Ray } from "../geometry/element/Ray.js";
import { Light } from "./Light.js";
export class SpotLight extends Light {
    public constructor(
        position: Vector3Dd,
        private readonly pointsAt: Vector3Dd,
        emission: ColorRgb,
        private readonly coefficient: number,
        private readonly radius: number,
        private readonly falloff: number,
    ) {
        super(position, emission);
    }
    public getPointsAt(): Vector3Dd {
        return this.pointsAt;
    }
    public getCoefficient(): number {
        return this.coefficient;
    }
    public getRadius(): number {
        return this.radius;
    }
    public getFalloff(): number {
        return this.falloff;
    }
    public override getDirectionAndDistance(point: Vector3Dd): Light.LightDirection {
        const v = this.getPosition().subtract(point),
            distance = v.length();
        if (distance <= VSDK.EPSILON) return new Light.LightDirection(new Vector3Dd(), 0);
        return new Light.LightDirection(v.multiply(1 / distance), distance - VSDK.EPSILON);
    }
    private static cubicSpline(low: number, high: number, pos: number): number {
        if (pos < low) return 0;
        if (pos > high) return 1;
        if (high === low) return 0;
        const t = (pos - low) / (high - low);
        return (3 - 2 * t) * t * t;
    }
    public override evaluateLightResponseFactor(ray: Ray): number {
        let d = this.pointsAt.subtract(this.getPosition());
        const length = d.length();
        if (length <= 0) return 0;
        d = d.multiply(1 / length);
        const cosine = -ray.getDirection().dotProduct(d);
        if (cosine <= 0) return 0;
        let response = Math.pow(cosine, this.coefficient);
        if (this.radius > 0) response *= SpotLight.cubicSpline(this.falloff, this.radius, cosine);
        return response;
    }
    public override copy(): Light {
        const copy = new SpotLight(
            this.getPosition(),
            this.pointsAt,
            this.getEmission(),
            this.coefficient,
            this.radius,
            this.falloff,
        );
        copy.setId(this.getId());
        copy.setName(this.getName());
        return copy;
    }
}
